src/planscape/forsys/cluster_stands.py

```diff
@@ -0,0 +1,261 @@
+        for x in pixel_dist_to_condition_values.keys():
+            if x not in normalized_pixel_dist_to_condition_values.keys():
+                normalized_pixel_dist_to_condition_values[x] = {}
+            for y in pixel_dist_to_condition_values[x].keys():
+                if y not in normalized_pixel_dist_to_condition_values[x].keys():
+                    normalized_pixel_dist_to_condition_values[x][y] = {}
+                conditions = pixel_dist_to_condition_values[x][y]
+                # Division by priority_condition_max_value forces individual
+                # condition values to be within range, [0, 1].
+                # A subsequent division by sqrt(len(conditions)) forces the l2
+                # norm of the condition vector to be within range, [0, 1].
+                denom = priority_condition_max_value * \
+                    np.sqrt(len(conditions.keys()))
+                for c in conditions.keys():
+                    normalized_pixel_dist_to_condition_values[x][y][c] = \
+                        conditions[c] / denom
+        return normalized_pixel_dist_to_condition_values
+
+    def _normalize_priority_weights(self,
+                                    priority_weights: dict[str, float]
+                                    ) -> dict[str, float]:
+        # The denominator is the l2 norm of the priority weights vector.
+        denom = 0
+        for p in priority_weights.keys():
+            denom = denom + np.square(priority_weights[p])
+        denom = np.sqrt(denom)
+
+        normalized_priority_weights = {}
+        for p in priority_weights.keys():
+            # Division by the l2 norm forces the l2 norm of the priority
+            # weights veector to be 1.
+            normalized_priority_weights[p] = priority_weights[p] / denom
+        return normalized_priority_weights
+
+    def _get_mask(
+        self, pixel_dist_to_condition_values: dict[int,
+                                                   dict[int,
+                                                        dict[str, float]]],
+            pixel_width: int, pixel_height: int) -> list[list[bool]]:
+        mask = np.full((pixel_width, pixel_height), False)
+        for x in range(pixel_width):
+            if x not in pixel_dist_to_condition_values.keys():
+                continue
+            for y in range(pixel_height):
+                if y not in pixel_dist_to_condition_values[x].keys():
+                    continue
+                mask[x][y] = True
+        return mask
+
+    def _get_features(self,
+                      pixel_dist_to_condition_values:
+                      dict[int, dict[int, dict[str, float]]],
+                      pixel_width: int, pixel_height: int,
+                      priority_weights: dict[str, float],
+                      pixel_index_weight: float) -> list[list[float]]:
+        features = []
+        for x in range(pixel_width):
+            if x not in pixel_dist_to_condition_values.keys():
```

Comment:
could this be "if mask[x][y]:"?

Reply:
yes: that might make the relationship between the list of feature vectors and the mask more clear, too.
Done.